x-pack/plugins/security_solution/public/detections/containers/detection_engine/rules/rules_table/use_rules_table.ts

```diff
@@ -0,0 +1,154 @@
+
+const INITIAL_SORT_FIELD = 'enabled';
+
+const initialStateDefaults: RulesTableState = {
+  rules: [],
+  pagination: {
+    page: 1,
+    perPage: 20,
+    total: 0,
+  },
+  filterOptions: {
+    filter: '',
+    sortField: INITIAL_SORT_FIELD,
+    sortOrder: 'desc',
+    tags: [],
+    showCustomRules: false,
+    showElasticRules: false,
+  },
+  loadingRulesAction: null,
+  loadingRuleIds: [],
+  selectedRuleIds: [],
+  exportRuleIds: [],
+  lastUpdated: 0,
+  isRefreshOn: true,
+  showIdleModal: false,
+};
+
+export interface UseRulesTableParams {
+  tableRef: React.MutableRefObject<EuiBasicTable<unknown> | undefined>;
+  initialStateOverride?: Partial<RulesTableState>;
+}
+
+export interface UseRulesTableReturn extends RulesTableFacade {
+  state: RulesTableState;
+  dispatch: Dispatch<RulesTableAction>;
+  reFetchRules: () => Promise<void>;
+}
+
+export const useRulesTable = (params: UseRulesTableParams): UseRulesTableReturn => {
+  const { tableRef, initialStateOverride } = params;
+
+  const initialState: RulesTableState = {
+    ...initialStateDefaults,
+    lastUpdated: Date.now(),
+    ...initialStateOverride,
+  };
+
+  const reducer = useMemo(() => createRulesTableReducer(tableRef), [tableRef]);
+  const [state, dispatch] = useReducer(reducer, initialState);
+  const facade = useRef(createRulesTableFacade(dispatch));
+
+  const reFetchRules = useRef<() => Promise<void>>(() => Promise.resolve());
+  const [, dispatchToaster] = useStateToaster();
+
+  // Fetch all rules at once (and only once) after mount. Filtering, sorting,
+  // searching and pagination are implemented in memory on the client side.
+  useEffect(() => {
+    let isSubscribed = true;
+    const abortCtrl = new AbortController();
+
+    const fetchData = async () => {
+      try {
+        facade.current.actionStarted('load', []);
+
+        const fetchRulesResult = await fetchRules({
+          filterOptions: initialStateDefaults.filterOptions,
+          pagination: {
+            page: 1,
+            perPage: NUM_RULES_TO_LOAD,
+            total: -1,
+          },
+          signal: abortCtrl.signal,
+        });
+
+        if (isSubscribed) {
+          facade.current.setRules(fetchRulesResult.data, {
+            total: fetchRulesResult.total,
+          });
+        }
+      } catch (error) {
+        if (isSubscribed) {
+          errorToToaster({ title: i18n.RULE_AND_TIMELINE_FETCH_FAILURE, error, dispatchToaster });
+          facade.current.setRules([], {});
+        }
+      }
+      if (isSubscribed) {
+        facade.current.actionStopped();
+      }
+    };
+
+    fetchData();
+    reFetchRules.current = () => fetchData();
+
+    return () => {
+      isSubscribed = false;
+      abortCtrl.abort();
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const projections = useProjectedState(state);
+
+  return {
+    state: { ...state, ...projections },
```

Comment:
Here we decorate state with projected values (derived state) which allows us to leave most of the code in components untouched.